srtcore/buffer.cpp

```diff
@@ -374,9 +372,11 @@ int CSndBuffer::readData(char** data, int32_t& msgno_bitset, uint64_t& srctime,
    }
    msgno_bitset = m_pCurrBlock->m_iMsgNoBitset;
 
-   srctime =
+   // TODO: Use source time if it is provided.
```

Comment:
Source time is a matter of FR #930 
I left those comments to be considered within the work for that FR. At the moment, source time is not proved to be working.
So I would kindly ask to leave those comments and commented out code.

Source time would only have been used if TSBPD is enabled, and if source time value is higher, than the start time of the connection. This is a huge source of potential issues.
```
 if (m_bPeerTsbPd)
 {
     if (origintime >= m_stats.tsStartTime)
         packet.m_iTimeStamp = count_microseconds(origintime - m_stats.tsStartTime);
     else
         packet.m_iTimeStamp = count_microseconds(steady_clock::now() - m_stats.tsStartTime);
 }
 else
 {
     packet.m_iTimeStamp = count_microseconds(steady_clock::now() - m_stats.tsStartTime);
 }
```

Reply:
I'd just add one thing: in order to rely on the user-provided timestamps, there must be some requirements put on this, these requirements must be verified, and also sending calls rejected if the requirements are not satisfied.

The source times must come from possibly two sources:

1. Another SRT connection that have provided these timestamps on the output.
2. Some live playing facility that has set up better times than the read-at-send-call times would be.

In the #1 case there's no problem - it's a passthrough. In case of #2 you need to put a requirement on it - like, for example, that the values are provided as MONOTONIC CLOCK - and this requirement should be then verified, and if:

 - the time is from too long past towards the currently read monotonic clock (like more than 1h)
 - the time is in the future towards the currently read monotonic clock

then the send calls should be rejected.

This is a nontrivial change and requires a little bit more work. It was probably working (not fully) before as it was relying on system clock, but this solution is vulnerable in case when your system clock can be reconfigured during transmission, therefore it should use monotonic clock - but then the monotonic clock must be also the base clock for the user-specified timestamps.

Therefore here this functionality be better turned off and the fix postponed.